lib/src/rules/avoid_implementing_value_types.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r"Don't implement classes that override `==`.";
+
+const _details = r'''**DON'T** implement classes that override `==`.
+
+The `==` operator is contractually required to be an equivalence relation;
+that is, symmetrically for all objects `o1` and `o2`, `o1 == o2` and `o2 == o1`
+must either both be true, or both be false.
+
+When using `implements`, you do not inherit the method body of `==`, making it
+nearly impossible to follow the contract of `==`. Classes that override `==`
+typically are usable directly in tests _without_ creating mocks or fakes as
+well.
+
+> _NOTE_: Dart does not have true _value types_, so instead we consider a class
+> that implements `==`  as a _proxy_ for identifying value types.
+
+For example, for a given class `Size`:
+
+```
+class Size {
+  final int inBytes;
+  const Size(this.inBytes);
+
+  @override
+  bool operator ==(Object other) => other is Size && other.inBytes == inBytes;
+
+  @override
+  int get hashCode => inBytes.hashCode;
+}
+```
+
+**BAD**:
+```
+class CustomSize implements Size {
+  final int inBytes;
+  const CustomSize(this.inBytes);
+
+  int get inKilobytes => inBytes ~/ 1000;
+}
+```
+
+**BAD**:
+```
+import 'package:test/test.dart';
+import 'size.dart';
+
+class FakeSize implements Size {
+  int inBytes = 0;
+}
+
+void main() {
+  test('should not throw on a size >1Kb', () {
+    expect(() => someFunction(new FakeSize()..inBytes = 1001), returnsNormally);
+  });
+}
+```
+
+**GOOD**:
+```
+class ExtendedSize extends Size {
+  ExtendedSize(int inBytes) : super(inBytes);
+
+  int get inKilobytes => inBytes ~/ 1000;
+}
+```
+
+**GOOD**:
+```
+import 'package:test/test.dart';
+import 'size.dart';
+
+void main() {
+  test('should not throw on a size >1Kb', () {
+    expect(() => someFunction(new Size(1001)), returnsNormally);
+  });
+}
+```
+
+''';
+
+class AvoidImplementingValueTypes extends LintRule implements NodeLintRule {
+  AvoidImplementingValueTypes()
+      : super(
+            name: 'avoid_implementing_value_types',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry) {
+    final visitor = new _Visitor(this);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  static bool _overridesEquals(
```

Comment:
Thanks, but I guess I should have been more clear. The method `lookUpConcreteMethod` will look in the target class and all of it's superclasses (whether introduced via `extends` or `with`). As currently implemented, this method will traverse the hierarchy multiple times. What I was proposing is that you could remove this method completely and replace it's invocation with `element.lookUpConcreteMethod('==', element.library) != null`. On further reflection, though, that's not quite right, because in this case it will always find the implementation in `Object`, so you'll need to check to see whether the result is from `Object`, rather than whether the result is `null`.

Reply:
Oh ok, thanks. I tried the following it does not work (does not lint what the old code linted):

```dart
  static bool _overridesEquals(ClassElement element) {
    for (final interface in element.interfaces) {
      final interfaceClass = interface.element;
      final equalsMethod = interfaceClass.lookUpConcreteMethod(
        '==',
        element.library,
      );
      if (equalsMethod != null) {
        final implementingType = equalsMethod.enclosingElement.type;
        return !implementingType.isObject;
      }
    }
    return false;
  }
```

Any ideas?

Reply:
Wow, I'm still failing to communicate clearly. Huge apologies! Try this:
```
static bool _overridesEquals(ClassElement element) => !element
    .lookUpConcreteMethod('==', element.library)
    ?.enclosingElement
    ?.type
    ?.isObject == false;
```
The invocation of `lookUpConcreteMethod` should always return the implementation of `==` from `Object` (unless there's a concrete implementation lower in the hierarchy), except in the case where there's a cycle in the type hierarchy, hence the null checks.

(I haven't actually run that code, so I might have made a mistake.)

Reply:
No worries. I'm still learning! 🔖 

Looks like it works! PTAL.